### DISCUSSION TITLE
patchwise train, val & test split (single subject available)

### DISCUSSION THREAD

**Discussion by MirenLurBarquin** [TorchIO-project/torchio#1079] [Q&A]
Hi all, 

I have a single subject available, and I would like to train and predict on it patchwise, without any transform. Is there any way I can get a train_patches_queue, val_patches_queue, and test_patches_queue that do not overlap?

Any other solution in TorchIO in which I don't need to split the image or save the patches is welcome. 

Thank you in advance!

**Comment by fepegar** [comment]
Hi, @MirenLurBarquin. Why would you need a queue for validation or testing?

**Reply by MirenLurBarquin** [reply]
Hi, @fepegar. Thank you for your answer. 

I am training patchwise from just one subject, therefore, for validation I used a queue following the example on the General [TorchIO_tutorial.ipynb](https://colab.research.google.com/github/fepegar/torchio-notebooks/blob/main/notebooks/TorchIO_tutorial.ipynb). Since I have a single image, I thought that using a queue in the testing would be helpful too. 

The approach I have right now is to create a customized grid sampler that subsets the possible patches into 'train', 'test', and 'validation,  before yielding a random patch from the corresponding subset. For that, first I had to split my initial volume in two halves (this does not imply a problem regarding the content of my sample) so that I can have more than one subject as input to `tio.SubjectsDataset()`, and then, to `tio.Queue()`. Since the possible patches don't change within the volume, it seems to work fine for now. Nevertheless, there is probably a better way to approach this matter.

**Reply by fepegar** [reply]
My suggestion would be to iterate over your validation and test samplers during the corresponding phases. The queue is probably useful for stochastic training only.

**Comment by romainVala** [comment]
@MirenLurBarquin  I would be currious to know what are the images your work with

My guess is that it is high resolution microscopy data

I have seen use case (with very high resolution 2D data) where this is indeed the first preprocessing step (generate from the input image patches for train / val / test )

As fernando says this is not currently handel with the torchio queue which has a different use (for 3D MRI)

But I do think this is an important point if we want to have use case of torchoi in the microscopy field

**Reply by MirenLurBarquin** [reply]
Hi @romainVala, 

I am working with a synchrotron X-ray nano-holotomography (XNH) 3D volume. The idea would be similar to the preprocessing step you mention, to be able to split the volume in test/val/test patch sets.

For now, I would say that the approach is working with a customized grid sampler that keeps the test/val/test patch sets disjoint (and defining sampler_train = sampler( ..., "train"), sampler_val = sampler( ..., "val"), sampler_test = sampler( ..., "test") that samples patches from the corresponding patch set)

**Reply by fepegar** [reply]
> the torchio queue which has a different use (for 3D MRI)

The queue can be used for anything, right? I think it should be fine in this context as well. But I'm not sure it makes sense for validation, where you should probably use a deterministic approach for consistency across epochs.

**Reply by MirenLurBarquin** [reply]
In the patch-based section in [TorchIO_tutorial.ipynb](https://colab.research.google.com/github/fepegar/torchio-notebooks/blob/main/notebooks/TorchIO_tutorial.ipynb) the data loaders train and validation are obtained as follows: 

```
training_batch_size = 32
validation_batch_size = 2 * training_batch_size

patch_size = 24
samples_per_volume = 5
max_queue_length = 300
sampler = tio.data.UniformSampler(patch_size)

patches_training_set = tio.Queue(subjects_dataset=training_set, max_length=max_queue_length, samples_per_volume=samples_per_volume, sampler=sampler, num_workers=num_workers, shuffle_subjects=True, shuffle_patches=True, )

patches_validation_set = tio.Queue( subjects_dataset=validation_set, max_length=max_queue_length, samples_per_volume=samples_per_volume, sampler=sampler, num_workers=num_workers, shuffle_subjects=False, shuffle_patches=False,)

training_loader_patches = torch.utils.data.DataLoader(patches_training_set, batch_size=training_batch_size)

validation_loader_patches = torch.utils.data.DataLoader(patches_validation_set, batch_size=validation_batch_size)
```

Then, in this case, the validation set is not fixed and changes from epoch to epoch even if shuffle_subjects = False and shufle_patches=False in the patches_validation_set, because the queue gets randomly refilled from epoch to epoch (or when the queue empties)?

**Reply by romainVala** [reply]
> > the torchio queue which has a different use (for 3D MRI)
> 
> The queue can be used for anything, right? I think it should be fine in this context as well. But I'm not sure it makes sense for validation, where you should probably use a deterministic approach for consistency across epochs.

Yes you are right for the patch extraction. I was thinking about the "all" process of applying transform and then extracting patches.
When it comes to transforms (especially for spatial transform) one want to first apply the transform to the all volume and then extract random patches. This fit well with 3D MRI because they are not too big. But with huge input data one can just not perform any transform on the all volume.
but ok this is an off topic comment (sorry)
